### PR TITLE
Fix backtraces in gdb (etc.) on AArch64 and ARM32

### PR DESCRIPTION
**Pull request by mshinwell**
The assembly veneer `caml_c_call` on AArch64 doesn't currently make a new frame, which appears to cause trouble in the debugger when unwinding:
```
(gdb) bt
#0  0x00000000004c0cd0 in caml_obj_dup ()
#1  0x00000000004c804c in caml_c_call ()
Backtrace stopped: previous frame identical to this frame (corrupt stack?)
(gdb)
```
This patch should fix the problem.  We now get:
```
(gdb) bt
#0  0x00000000004c0cd0 in caml_obj_dup ()
#1  0x00000000004c8054 in caml_c_call ()
#2  0x0000000000482280 in camlRandom__entry () at random.ml:151
#3  0x00000000004432a4 in caml_program ()
#4  0x00000000004c80e8 in caml_start_program ()
#5  0x00000000004ae4cc in caml_startup_common ()
#6  0x00000000004ae518 in caml_startup ()
#7  0x0000000000442e38 in main ()
(gdb)
```
@xavierleroy Does this look correct to you?

### CODE REVIEW

**Comment by mshinwell** [comment]
Hang fire, I think this has a bug.  I will deal with it tomorrow.

**Comment by xavierleroy** [comment]
It makes me grumpy that we have to add two memory accesses to a hand-optimized and performance-critical piece of code just so that gdb doesn't lose its marbles.  

Could the problem with the current code be due to the fact that a return address is held in a nonstandard location?  (register x19).  I'm pretty sure that fact can be expressed with CFI declarations.

**Comment by mshinwell** [comment]
Good idea, that was needed.  However actually using any CFI directives at all would have been useful too; there was a missing `#include` !

I also added a CFI directive such that backtraces start neatly at `caml_start_program`.  OK?

**Comment by mshinwell** [comment]
Actually, hmm, maybe we should have the backtrace right back to `main`.  Let me try that.

**Comment by mshinwell** [comment]
Yeah, we should have the backtrace right through, otherwise `caml_callback` would truncate backtraces.  I think the current code works correctly now:
```
(gdb) bt
#0  0x000000000041e34c in caml_obj_dup ()
#1  0x0000000000424754 in caml_c_call ()
#2  0x000000000040e070 in camlRandom__entry () at random.ml:151
#3  0x00000000004021a4 in caml_program ()
#4  0x00000000004247e4 in caml_start_program ()
#5  0x0000000000424f54 in caml_startup_common ()
#6  0x0000000000424fa0 in caml_startup ()
#7  0x0000000000401ee8 in main ()
(gdb)
```

File: asmrun/arm64.S [hunk 1]
```diff
@@ -16,6 +16,8 @@
 /* Asm part of the runtime system, ARM processor, 64-bit mode */
 /* Must be preprocessed by cpp */
 
+#include "caml/m.h"
+
```
Comment (xavierleroy) [inline asmrun/arm64.S:20]: Regrettable omission indeed!

File: asmrun/arm64.S [hunk 2]
```diff
@@ -296,6 +302,7 @@ caml_c_call:
         PROFILE
     /* Preserve return address in callee-save register x19 */
         mov     x19, x30
+        CFI_REGISTER(30, 19)
     /* Record lowest stack address and return address */
```
Comment (xavierleroy) [inline asmrun/arm64.S:308]: Lovely!

**Review by xavierleroy — APPROVED** [review]
This is great!  I bow to your mastery of CFI directives!

**Comment by xavierleroy** [comment]
Feel free to squash-and-merge when you're ready, or tell me when you're happy with this PR and I'll squash-and-merge myself.

**Comment by gasche** [comment]
Wouldn't it make sense to have the same logic in `arm.S`? I looked for other architectures that (1) save the return address in a callee-save register and (2) have some sort of CFI support already, and found only `power`, which already has the corresponding CFI directive.

**Comment by xavierleroy** [comment]
It would make sense.  It would also make sense to add the missing `#include "m.h"` in arm.S as well :-)

**Comment by mshinwell** [comment]
I've made the suggested modifications to `arm.S`, but I'm currently testing it, so don't merge yet.

I also noticed on AArch64 that unwinding through `caml_call_gc` didn't work (same reason as for `caml_start_program`).  I have fixed that.

**Comment by xavierleroy** [comment]
The `arm.S` changes look good, but remember to mention them in the Changes file along with the `arm64.S` changes.

**Comment by xavierleroy** [comment]
Looks just right, thanks!  Squashed and merged manually to resolve the spurious conflict on Changes. (Commit  1501bc9 .)

**Comment by mshinwell** [comment]
Sorry, this wasn't ready -- we're still fixing bugs on ARM32.  Should be done tomorrow am.

**Comment by mshinwell** [comment]
I've reverted yesterday's commit.  Will ping again here when the final version is ready.

**Comment by mshinwell** [comment]
This has now been fixed properly for ARM32.  Note that this needs #1338 to function correctly though.

@xavierleroy Please review, squash and merge at your earliest convenience, with apologies for wasted merging work yesterday.

**Comment by xavierleroy** [comment]
Any progress on this PR?

**Comment by mshinwell** [comment]
Oh, sorry -- see last comment above.  It's ready for your review, thanks.

**Comment by xavierleroy** [comment]
Looks fine.  Merging right now.

**Comment by xavierleroy** [comment]
Cherry-picked as [4.06 8b4f260].

**Pull request merged**